### PR TITLE
Unsafe Code review

### PR DESCRIPTION
**Pull request by Plecra**
I thought I'd start by quickly looking through the project's unsafe code. Found one bug, and made a few changes that should improve readability for rustaceans 😁 

Matching bytes to enums will compile to the same code as a transmute. It *is* a bit of a pain, and it might be nice to use a crate like [`num-derive`](https://docs.rs/num-derive/) to automate it.

### CODE REVIEW

**Comment by slightknack** [comment]
Thanks for taking the time to put in this PR! I'm glad you cleaned up `tagged.rs`, and think the changes are a great boon for readability and safety.¹

However, I'm not a huge fan of switching `mem::transmute` to a large match statement:

1. In the parser, `Token::End` is *guaranteed* to always be the token with highest precedence. Because we check this explicitly, this expression is safe. Because the parser is not set it stone, and we'll be adding new precedence levels, it's nicer to have things all in one place. For this reason, I think that we should keep this as it is and just explicitly explain the safety instead.

2. My one concern with using a 

<details>
<summary>
<a href="https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&code=use%20std%3A%3Amem%3B%0A%0A%23%5Brepr(u8)%5D%0A%23%5Bderive(Debug)%5D%0Aenum%20Switch%20%7B%0A%20%20%20%20Zero%20%3D%200%2C%0A%20%20%20%20One%2C%0A%20%20%20%20Two%2C%0A%20%20%20%20Three%2C%0A%20%20%20%20Four%2C%0A%20%20%20%20Five%2C%0A%20%20%20%20Six%2C%0A%7D%0A%0A%0Afn%20from_byte_transmute(byte%3A%20u8)%20-%3E%20Switch%20%7B%0A%20%20%20%20unsafe%20%7B%20std%3A%3Amem%3A%3Atransmute(byte)%20%7D%0A%7D%0A%0Afn%20from_byte_match(byte%3A%20u8)%20-%3E%20Switch%20%7B%0A%20%20%20%20use%20Switch%3A%3A*%3B%0A%20%20%20%20%0A%20%20%20%20match%20byte%20%7B%0A%20%20%20%20%20%20%20%200%20%3D%3E%20Zero%2C%0A%20%20%20%20%20%20%20%201%20%3D%3E%20One%2C%0A%20%20%20%20%20%20%20%202%20%3D%3E%20Two%2C%0A%20%20%20%20%20%20%20%203%20%3D%3E%20Three%2C%0A%20%20%20%20%20%20%20%204%20%3D%3E%20Four%2C%0A%20%20%20%20%20%20%20%205%20%3D%3E%20Five%2C%0A%20%20%20%20%20%20%20%206%20%3D%3E%20Six%2C%0A%20%20%20%20%20%20%20%20_%20%3D%3E%20unreachable!()%2C%0A%20%20%20%20%7D%0A%7D%0A%0Afn%20main()%20%7B%0A%20%20%20%20let%20value_1%20%3D%20from_byte_transmute(3u8)%3B%0A%20%20%20%20let%20value_2%20%3D%20from_byte_match(3u8)%3B%0A%7D">Here's the generated assembly</a> for a toy example:
</summary>
<pre>
; with a transmutation
playground::from_byte_transmute:
	subq	$3, %rsp
	movb	%dil, 1(%rsp)
	movb	%dil, 2(%rsp)
	movb	2(%rsp), %al
	movb	%al, (%rsp)
	movb	(%rsp), %al
	addq	$3, %rsp
	retq
</pre>
<pre>
; with a match statement
playground::from_byte_match:
	subq	$24, %rsp
	movb	%dil, 23(%rsp)
	movzbl	%dil, %eax
	movl	%eax, %ecx
	subb	$6, %dil
	movq	%rcx, 8(%rsp)
	ja	.LBB12_1
	leaq	.LJTI12_0(%rip), %rax
	movq	8(%rsp), %rcx
	movslq	(%rax,%rcx,4), %rdx
	addq	%rax, %rdx
	jmpq	*%rdx
</pre>
Note that this is a debug build, because the Rust compiler inlines these functions otherwise.
</details>

Sensibly, the match statement has an added bounds check. This, however, is quite literally _The_ Hot Path: I'd prefer we skipped the bounds check and just transmuted instead - when optimized, this is a no-op.

If you're worried about the correctness of bytecode, i.e. that incorrect bytecode will never trigger a panic due to transmutation during runtime, we can do one of two things:

1. Add a verification step that walks the code and checks bounds before running, similar to the one that exists to debug/display `Lambda`.
2. Or, alternatively, defer the bounds check to runtime.

The VM assumes the generated bytecode to be correct, which allows for optimization; this will only become more true as we're able to verify more things at compile time.

> Matching bytes to enums will compile to the same code as a transmute. It is a bit of a pain, and it might be nice to use a crate like num-derive to automate it.

I'll take a look at that crate, but Passerine is no dep. If the core macro is simple enough, we can reimplement it for reuse in Passerine.

**TL;DR:** The tagged pointer is a welcome cleanup, but I'm unsure about the lexer and opcode. If you remove the lexer and opcode cleanups, I will merge now; we can continue discussing the trade-offs of the other two later.

**Comment by slightknack** [comment]
An additional note: Passerine assumes the layout of pointers to be 64-bits with the 16 high bits unset. This is true of most modern architectures, but not always. We should manually specify the pointer layout, and add an option to opt out of NaN-tagging.

**Comment by Plecra** [comment]
> If you're worried about the correctness of bytecode, i.e. that incorrect bytecode will never trigger a memory error due to transmutation during runtime, we can add a verification step that walks the code and checks bounds before running, similar to the one that exists to debug/display `Lambda`.

This sounds good! I'm interested in minimising unsafe code, and this would allow us to write a quick implementation of

```rust
fn to_bytecode(bytes: &[u8]) -> &[OpCode];
```

w.r.t the macros, I think it's a good idea to have a `macros` crate to tidy up a few patterns that'd normally be taken from external crates. I'll have a look to see if there are other patterns proc macros would help with. (It's possible to write as a declarative macro too, but it's grotesque :D)

**Comment by slightknack** [comment]
About the `to_bytecode` function:
> ```rust
> fn to_bytecode(bytes: &[u8]) -> &[OpCode];
> ```

Bytecode is represented as a series of bytes:
```rust
[0x_1, 0x_2, ..., Ox_N]
```
Each byte is one of two things: 

1. An `Opcode`
2. A numeric argument

Each Opcode takes N arguments, for example, the `Opcode` to `Load` a variable takes 1 numeric constant, the index of the variable on the stack:
```rust
// load x:
[Opcode::Load, x]
// Load is 0x08
// x is index 2 on stack
// bytecode is
[0x08, 0x02]
```

Numeric arguments are arbitrary numbers, split into bytes:
```
[0b1______1, 0b1______2, ..., 0b0______N]
```
Each chunk of 7 bits is spliced into a byte; the first bit in each byte is the continue flag; the number is done when the continue flag is set to zero. You can find the code that splits and reconstructs numbers here:

https://github.com/vrtbl/passerine/blob/1f81a1c296b79a4db1d7837f6640a16104a0a795/src/common/number.rs#L10

To statically verify bytecode, we:
1. Look at the next byte
2. Determine whether it is a valid opcode
3. Determine the number of arguments it takes
4. Reconstruct the numeric arguments
5. Move onto the next byte

If we reach an opcode that is not valid, or end in the middle of decoding numeric argument(s), the bytecode is invalid. So, back to this function:
> ```rust
> fn to_bytecode(bytes: &[u8]) -> &[OpCode];
> ```

We can't implement this function because bytecode is more than just a string of opcodes, if that makes sense; it's a series of Opcodes and numeric arguments.

> w.r.t the macros, I think it's a good idea to have a `macros` crate to tidy up a few patterns that'd normally be taken from external crates. I'll have a look to see if there are other patterns proc macros would help with. (It's possible to write as a declarative macro too, but it's grotesque :D)

Sounds good; something I've been wanting to do for a while is to implement a serde-style derive macro that maps Rust structs to Passerine Data. I'm not sure how to do this, so if you have any insights please let me know. Thanks again!

**Comment by slightknack** [comment]
By the way, I'm going to cherry-pick and merge the commit you sent that refactors tagged data.

**Comment by Plecra** [comment]
Continuing the refactor of transmutes in #40
